### ISSUE TITLE
Duplicated child RushObjects with circular relationships

### ISSUE THREAD

**Issue by PaulineBAUDOUIN** [Stuart-campbell/RushOrm#29]
Hi, it's me again, I have a problem which I can't manage to solve...
I have a first ActivityA which starts ActivityB when onClick on a button.
In my ActivityB, I want to create a new Projet (RushObject) so here is what I do to create it:
``` Java
Projet toCreateProj = new Projet();
            toCreateProj.identifier = 1;
            toCreateProj.title = editText.getText().toString();
            toCreateProj.maquette = selectedMaq; 
            toCreateProj.user = utilisateur;
            toCreateProj.save();
            utilisateur.projets.add(toCreateProj); // a Utilisateur (RushObject) has many Projets
            utilisateur.save();
            System.out.println(toCreateProj.maquette.gabarits.size()); // returns 3
            dialog.cancel();
            finish();
```
selectedMaquette is a Maquette (RushObject) which I have by making a RushSearch. Same for utilisateur. **My System.out returns 3 which is normal.** 
On finish(), I go back to my ActivityA. In the onResume() i do:
``` Java
Projet toBeTested = new RushSearch().whereEqual("identifier", 1).findSingle(Projet.class); // I am sure it is the good project
        if(toBeTested != null){
            System.out.println("BIDON " + toBeTested.maquette.gabarits.size()); // returns 6
        }
```
**But now, my System.out returns 6** but should still return 3. I really don't get why... Any idea ? 
Thanks

**Comment by Stuart-campbell** [comment]
You might have multiple projects with identifier = 1 that you create will developing? So find single might not be returning the object you just created?

**Comment by PaulineBAUDOUIN** [comment]
In my whole application, I only create one Projet so it can only be this one I think. I am really sure this is the Projet I want !

**Comment by Stuart-campbell** [comment]
Try removing this line.
  
    toCreateProj.save();

As saving utilisateur should save all the children.

Although wondering if something might be going wrong saving children that also have children.

**Comment by PaulineBAUDOUIN** [comment]
That's also what I thought because in my application, there are RushObjects:
- Maquette associated with one or many Utilisateurs and Projets.
-  Utilisateur associated with one or many Maquettes and Projets.
- Projet associated with one Maquette and one Utilisateur.

So there is a triangle which might be the cause of the duplication.

**Comment by Stuart-campbell** [comment]
Does this look like your example.

    ClassA extends RushObject {
        List<ClassB> classBs
        List<ClassC> classCs
     }

    ClassB extends RushObject {
        List<ClassA> classAs;
        List<ClassC> classCs;
    }

    ClassC extends RushObject {
        ClassA classA;
        ClassB classB;
    }

If it is ill try to reproduce. Might take a little time as not the most simple example.

**Comment by PaulineBAUDOUIN** [comment]
Ok so I changes a little bit the structure of the database into a more simple one so here is the new one:

``` Java
Gabarit extends RushObject{
}

Maquette extends RushObject{
    List<Gabarit> gabarits;
}

Utilisateur extends RushObject{
    List<Maquette> maquettes;
}

Projet extends RushObject{
    Maquette maquette;
    Utilisateur user;
}
```
So when I create a new Projet as I do in my first post, it duplicates all the items in the list gabarits of the Maquette class and I don't understand why...

**Comment by PaulineBAUDOUIN** [comment]
Here is what I do to save a Projet:
```Java
Maquette m1 = new RushSearch().whereEqual("nom", selectedMaq.nom).findSingle(Maquette.class);
            System.out.println("size 1: " + m1.gabarits.size()); // displays 3
            Projet toCreateProj = new Projet();
            toCreateProj.identifier = 1;//toCreateProj.getNewIdentifier();
            toCreateProj.title = editText.getText().toString();
            toCreateProj.maquette = selectedMaq;
            toCreateProj.user = utilisateur;
            toCreateProj.save();

            Maquette m2 = new RushSearch().whereEqual("nom", selectedMaq.nom).findSingle(Maquette.class);
            System.out.println("size 2: " + m2.gabarits.size()); // displays 6
```

So my first System.out displays 3 (normal) and the second one displays 6 because of duplication

**Comment by Stuart-campbell** [comment]
From what you said looks like the join table might not be clearing. The tests are a little thin on chained relationships. I look into this in detail asap. Thanks for reporting it.

**Comment by PaulineBAUDOUIN** [comment]
I managed to fix my problem by:
- creating a new RushObject called Autorisation wich contains a Utilisateur and a List of Maquettes
- removing the list of Maquettes in the Utilisateur RushObject 

which makes that all the relations go only in one way !
It worked for me

**Comment by Stuart-campbell** [comment]
So that confirms the bug is in the circular relationship.

**Comment by PaulineBAUDOUIN** [comment]
Yes indeed, thank you for helping me !

**Comment by HitOdessit** [comment]
@Stuart-campbell, any updates regarding fixing this bug?

**Comment by Stuart-campbell** [comment]
@HitOdessit been a little busy last week or so. Did find the time to put together a number of unit tests to attempt to pin point the bug with no luck.

Would you be able to share your data model in the most basic form and the order in which you save things see if I can get my tests to fail.

Sorry for the delay mix of weddings and release deadlines.

Thanks

**Comment by HitOdessit** [comment]
@Stuart-campbell, sure, no problems, will provide you samples later today.

**Comment by HitOdessit** [comment]
@Stuart-campbell  Please look at the following project which can reproduce the error: https://github.com/nmoskalenko/rushorm_testing
We've tried to minimize amount of code, and also added some commets to help you understand the problem.

**Comment by Stuart-campbell** [comment]
@HitOdessit 

Can you have a look at this unit test. As far as I can tell its identical to the project you created yet passes on api 14 and api 21 nexus 4 + 5

https://github.com/Stuart-campbell/RushOrm/blob/master/RushORM/app/src/androidTest/java/co/uk/rushexample/BugTests.java#L85-L113

Mapping of object names from the example 

Bug29A -> E

https://github.com/Stuart-campbell/RushOrm/blob/master/RushORM/app/src/main/java/co/uk/rushexample/testobjects/Bug29A.java

Bug29B -> C

https://github.com/Stuart-campbell/RushOrm/blob/master/RushORM/app/src/main/java/co/uk/rushexample/testobjects/Bug29B.java

Bug29C -> P

https://github.com/Stuart-campbell/RushOrm/blob/master/RushORM/app/src/main/java/co/uk/rushexample/testobjects/Bug29C.java

I coming to a bit of a dead end on this one. Might be worth you pulling the project and trying to take a look yourself see if you can point me in the right direction.

Thanks

**Comment by nmoskalenko** [comment]
@Stuart-campbell ,

the issue only partially fixed in 1.1.8 

please look into update of the sample here - https://github.com/nmoskalenko/rushorm_testing
in you test, please check the Bug29A's list size after the test. 
It's size expected to be 2, but actually 4.

**Comment by Stuart-campbell** [comment]
Ah! I can not reproduce it and I finally think I know what could be the issue.

Should get it fixed tonight.

Thanks

**Comment by Stuart-campbell** [comment]
Fixed in v1.0.9

**Comment by nmoskalenko** [comment]
Just tested and can confirm fix in 1.1.9
Thanks Stuart!

**Comment by aercanozcan** [comment]
awesome!
